cmake/modules/yaml.cmake

```diff
@@ -0,0 +1,403 @@
+  zephyr_check_arguments_exclusive(${CMAKE_CURRENT_FUNCTION} ARG_YAML VALUE LIST)
+  internal_yaml_context_required(NAME ${ARG_YAML_NAME})
+
+  get_property(json_content TARGET ${ARG_YAML_NAME}_scope PROPERTY JSON)
+
+  set(yaml_key_undefined ${ARG_YAML_KEY})
+  foreach(k ${yaml_key_undefined})
+    list(REMOVE_AT yaml_key_undefined 0)
+    # We ignore any errors as we are checking for existence of the key, and
+    # non-existing keys will throw errors but also set type to NOT-FOUND.
+    string(JSON type ERROR_VARIABLE ignore TYPE "${json_content}" ${valid_keys} ${k})
+
+    if(NOT type)
+      list(APPEND yaml_key_create ${k})
+      break()
+    endif()
+    list(APPEND valid_keys ${k})
+  endforeach()
+
+  list(REVERSE yaml_key_undefined)
+  if(NOT "${yaml_key_undefined}" STREQUAL "")
+    if(ARG_YAML_APPEND)
+      set(json_string "[]")
+    else()
+      set(json_string "\"\"")
+    endif()
+
+    foreach(k ${yaml_key_undefined})
+      set(json_string "{\"${k}\": ${json_string}}")
+    endforeach()
+    string(JSON json_content SET "${json_content}"
+           ${valid_keys} ${yaml_key_create} "${json_string}"
+    )
+  endif()
+
+  if(DEFINED ARG_YAML_LIST)
+    if(NOT ARG_YAML_APPEND)
+      string(JSON json_content SET "${json_content}" ${ARG_YAML_KEY} "[]")
+    endif()
+
+    string(JSON subjson GET "${json_content}" ${ARG_YAML_KEY})
+    string(JSON index LENGTH "${subjson}")
+    list(LENGTH ARG_YAML_LIST length)
+    math(EXPR stop "${index} + ${length} - 1")
+    foreach(i RANGE ${index} ${stop})
+      list(POP_FRONT ARG_YAML_LIST value)
+      string(JSON json_content SET "${json_content}" ${ARG_YAML_KEY} ${i} "\"${value}\"")
+    endforeach()
+  else()
+    string(JSON json_content SET "${json_content}" ${ARG_YAML_KEY} "\"${ARG_YAML_VALUE}\"")
+  endif()
+
+  zephyr_set(JSON "${json_content}" SCOPE ${ARG_YAML_NAME})
+endfunction()
+
+# Usage:
+#   yaml_remove(NAME <name> KEY <key>...)
+#
+# Remove the KEY <key>... from the YAML context <name>.
+#
+# Several levels of keys can be given, for example:
+# KEY build cmake command
+#
+# To remove the key 'command' underneath 'cmake' in the toplevel 'build'
+#
+# NAME <name>: Name of the YAML context.
+# KEY <key>  : Name of key to remove.
+#
+function(yaml_remove)
+  cmake_parse_arguments(ARG_YAML "" "NAME" "KEY" ${ARGN})
+
+  zephyr_check_arguments_required_all(${CMAKE_CURRENT_FUNCTION} ARG_YAML NAME KEY)
+  internal_yaml_context_required(NAME ${ARG_YAML_NAME})
+
+  get_property(json_content TARGET ${ARG_YAML_NAME}_scope PROPERTY JSON)
+  string(JSON json_content REMOVE "${json_content}" ${ARG_YAML_KEY})
+
+  zephyr_set(JSON "${json_content}" SCOPE ${ARG_YAML_NAME})
+endfunction()
+
+# Usage:
+#   yaml_write(NAME <name>)
+#
+# Write the YAML context <name> to the file which were given with the earlier
+# 'yaml_load()' or 'yaml_create()' call.
+#
+# NAME <name>: Name of the YAML context
+#
+function(yaml_write)
```

Comment:
Wouldn't it make sense to decouple the context from the file? And add a `yaml_save` function that can be passed a filename?

Reply:
I guess it depends on the use-case.
The two use-cases I currently had in mind were
1. you know everything at one location, like the [domains.yaml](https://github.com/zephyrproject-rtos/zephyr/blob/main/share/sysbuild/cmake/domains.cmake) file.
    There it doesn't really matter.
2. You may know about the file (and load it into a context) at one location, but you'll be updating the context at other locations, then all you want to know is the agreed context name so that you can. You don't want/need to know about file names. That's why the actual filename is a property on the context.

but I see your proposal / use-case as just as valid, so perhaps allowing for either write to specific filename, or omitting the filename when creating a new context.

In short:
- allow specifying a file name on `yaml_write()`
  The filename property of the context will be used when filename is not provided with `yaml_write`.
    - Fail with error if calling `yaml_write()` if context has no filename property defined and filename was not passed as argument.
- make filename optional on `yaml_create()`

Reply:
I think these additions are welcome, can we rename `yaml_write` to `yaml_save`? If you use `load`, then `save` seems better suited.

Would it also make sense to allow loading multiple files to amend/overwrite values in the current context? For this the decoupling makes over more sense.

Reply:
> can we rename yaml_write to yaml_save? If you use load, then save seems better suited.

Good point and I agree, will do.

> Would it also make sense to allow loading multiple files to amend/overwrite values in the current context? For this the decoupling makes over more sense.

I would leave that for a future enhancement.
But one use-case (not yet supported) which I consider useful, is to be able to create a new context and then merge that into a context loaded from file. That principle is very similar to the loading of multiple files.

Reply:
> I think these additions are welcome, can we rename yaml_write to yaml_save?

Done